### PR TITLE
Move playlist and music track code from Audio into its own file & namespace, Jukebox

### PR DESCRIPTION
**Pull request by LeeSpork**
Completes #2763

Also increased the usage of `MusicId` (i.e. in place of `uint8_t`) in both the existing file and the new file.

### CODE REVIEW

File: src/OpenLoco/src/Jukebox.cpp
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Jukebox.cpp:16]: We no longer use vanilla routines for any audio, so these variables can actually be integrated!
```suggestion
    static uint8_t _currentSong; // 0x0050D434
    static uint8_t _lastSong;    // 0x0050D435
```

File: src/OpenLoco/src/Jukebox.cpp
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Jukebox.cpp:112]: This is only used on one line, correct? I don't think it's worth it tbh.

File: src/OpenLoco/src/Jukebox.cpp
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Jukebox.cpp:11]: Interop can be removed once the two `loco_global`s have been integrated, i.e.:
```suggestion
#include <numeric>

using namespace OpenLoco::Environment;
```

File: src/OpenLoco/src/Jukebox.cpp
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Jukebox.cpp:147]: I think we might as well move to a const ref for these. What do you think?

File: src/OpenLoco/src/Jukebox.cpp
Comment (LeeSpork) [inline src/OpenLoco/src/Jukebox.cpp:121]: This should be impossible to happen, as the music selection window does not allow you to select 0 tracks for the 'customised selection' mode (and you have bigger problems if the 'all music' comes up empty). Should I remove this?

Comment (LeeSpork) [reply src/OpenLoco/src/Jukebox.cpp]: There is another check right bellow this that adds all music if the playlist is (still) empty, which could in theory happen for currentEra mode, maybe. Though I'm also not sure if that is the best way to deal with there being no current era music?

**Comment by LeeSpork** [comment]
The remaining music-related functions in Audio.cpp are (by my count):
- `void revalidateCurrentTrack()`
- `void playBackgroundMusic()`
- `static std::optional<uint32_t> loadMusicSample(PathId asset)`
- `void playMusic(PathId sample, int32_t volume, bool loop)`
- `void resetMusic()`
- `void stopMusic()`

File: src/OpenLoco/src/Audio/Audio.cpp
Comment (LeeSpork) [inline src/OpenLoco/src/Audio/Audio.cpp:55]: Would it make sense to use `~0` instead of `0xFF` here?

File: src/OpenLoco/src/Jukebox.cpp
Comment (LeeSpork) [reply src/OpenLoco/src/Jukebox.cpp:16]: _lastSong is now used in both Audio.cpp and Jukebox.cpp, but I change that.
Should I apply this in Audio.cpp, or should that be done in a separate pull request?

**Comment by LeeSpork** [comment]
I'm still not feeling entirely sure about the name of the new files & namespace, but I don't think it is a bad name for it.

File: src/OpenLoco/src/Audio/Audio.cpp
Comment (LeeSpork) [reply src/OpenLoco/src/Audio/Audio.cpp:55]: Another alternative is `UCHAR_MAX`, though that also wouldn't change if MusicId is redefined.

Comment (AaronVanGeffen) [reply src/OpenLoco/src/Audio/Audio.cpp:55]: Rather than using C constants, you could use std::limits instead. Not sure it's worth it, though.

**Comment by RedMarcher** [comment]
I noticed that you just copied over the old logic into jukebox.cpp. That's fine by me, but I'd definitely like to make some adjustments since I think the current way of selecting songs is wasteful and unintuitive.

I think its a good name. I'd just merge this as is and I will add my changes to a different pull request.

**Comment by RedMarcher** [comment]
By any chance, is the name music.cpp already taken? I think that might be a more descriptive name than jukebox

File: src/OpenLoco/src/Audio/Audio.cpp
Comment (LeftofZen) [reply src/OpenLoco/src/Audio/Audio.cpp:55]: I would just use the standard 0xFF since that's what the rest of the codebase uses

File: src/OpenLoco/src/Jukebox.cpp
Comment (LeftofZen) [reply src/OpenLoco/src/Jukebox.cpp]: if its part of the refactor/move, then yeah you can apply it here as well

File: src/OpenLoco/src/Jukebox.cpp
Comment (LeftofZen) [reply src/OpenLoco/src/Jukebox.cpp]: if no music for any playlist, there is nothing to do - no music should play

**Comment by LeftofZen** [comment]
> but I'd definitely like to make some adjustments since I think the current way of selecting songs is wasteful and unintuitive.

LeeSpork has already refactored this and the code is fine. Also, it is bad practice to include multiple changes in single PRs. This is just a refactor PR. You can make features/functional changes in a separate PR.

Music.cpp is a *less* descriptive name than something like Jukebox.cpp or Playlist.cpp since Music is a generic term, but Playlist.cpc describes what is actually happening in terms of functionality.

**Comment by RedMarcher** [comment]
> > but I'd definitely like to make some adjustments since I think the current way of selecting songs is wasteful and unintuitive.
> 
> LeeSpork has already refactored this and the code is fine. If you have specific suggestions, please feel free to suggest or make a PR. Also, it is bad practice to include multiple changes in single PRs. This is just a refactor PR. You can make features/functional changes in a separate PR.
> 
> Music.cpp is a _less_ descriptive name than something like Jukebox.cpp or Playlist.cpp since Music is a generic term, but Playlist.cpc describes what is actually happening in terms of functionality.

Yeah, I think you're right about the name. Well, I will wait until this is merged before doing anything else. It's good that this was moved to a different file, since the original was definitely getting too much feature overlap.

File: src/OpenLoco/src/Jukebox.cpp
Comment (RedMarcher) [reply src/OpenLoco/src/Jukebox.cpp]: Don't worry too much about the logic, I would like to rewrite most of it anyways once this is merged. Thanks for making the new cpp and h files.

**Review by RedMarcher — APPROVED** [review]
Looks like it maintains the original logic. That's more than fine. This needs to be merged first before any changing of the original logic.

**Review by LeftofZen — APPROVED** [review]
playtested, seems to work as  it currently does.

I found that the same song can still play twice for some reason, but this is the current (24.11) behaviour anyways, so nothing has changed

File: src/OpenLoco/src/Audio/Audio.cpp
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Audio/Audio.cpp:1014]: Oops, I think this was making a copy. Could you change this to:
```suggestion
                auto& info = Jukebox::getMusicInfo(_currentSong);
```

File: src/OpenLoco/src/Jukebox.cpp
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Jukebox.cpp:89]: I think this is creating a new symbol. Could you try if this works instead:
```suggestion
        using Config::MusicPlaylistType;
```

**Review by AaronVanGeffen — COMMENTED** [review]
Two small suggestions, LGTM otherwise!

**Comment by LeeSpork** [comment]
> I found that the same song can still play twice for some reason, but this is the current (24.11) behaviour anyways, so nothing has changed

Yes, changing the behaviour was outside of my goals for this pull. If you are referring to the ⏭ button (#2703), I wasn't planning on thinking about that issue until this after this pull has been merged.

Also note that the same song technically can't play twice in a row (which #2750 will fix if it is merged). (⏭ apparently happens to avoid this by [playing "None" for one frame](https://github.com/OpenLoco/OpenLoco/issues/2704))

File: src/OpenLoco/src/Jukebox.cpp
Comment (LeeSpork) [reply src/OpenLoco/src/Jukebox.cpp:89]: Seems to work 👍

**Pull request merged**
